src/ILCompiler.TypeSystem/tests/GenericTypeAndMethodTests.cs

```diff
@@ -0,0 +1,56 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Linq;
+using Internal.TypeSystem;
+using Xunit;
+
+namespace TypeSystemTests
+{
+    public class GenericMethodTests
+    {
+        private TestTypeSystemContext _context;
+        private ModuleDesc _testModule;
+
+        public GenericMethodTests()
+        {
+            _context = new TestTypeSystemContext(TargetArchitecture.Unknown);
+            var systemModule = _context.CreateModuleForSimpleName("CoreTestAssembly");
+            _context.SetSystemModule(systemModule);
+
+            _testModule = systemModule;
+        }
+
+        /// <summary>
+        /// Testing proper instantiation of methods involving generic parameters in their signature.
+        /// </summary>
+        [Fact]
+        [ActiveIssue(-1)]
+        public void TestMethodInstantiation()
+        {
+            MetadataType t = _testModule.GetType("GenericTypes", "GenericClass`1");
+            MethodDesc fooMethod = t.GetMethods().First(m => m.Name == "Foo");
+            MethodDesc instantiatedMethod =
```

Comment:
It's about the difference between doing InstantiateSignature on a generic definition vs on a generic instantiated over it's own formals.

Let's say I have a method body within a generic method `SomeMethod<T>` that does a ldtoken of itself in two ways:
```
ldtoken SomeType.SomeMethod
...
ldtoken SomeType.SomeMethod<!!T>
```

The first one loads the definition, the second one loads instantiation over it's own formal.

When we do codegen for this method, we call `InstantiateSignature` on everything within the body. The first one needs to stay a definition even after substitutions.